### ISSUE TITLE
Accept additional string formatting styles in sankey

### ISSUE THREAD

**Issue by CharlesHe16** [matplotlib/matplotlib#19133]
Currently, Sankey class accepts number formatting for labels using the old printf string formatting. 

I think the relevant line is here:

https://github.com/matplotlib/matplotlib/blob/a02ffbcd898793c95de66d3bdb4dc0a8b6c5a3d0/lib/matplotlib/sankey.py#L742

F-strings seem to be preferred over this format (you probably know this already but here's [some reasons](https://realpython.com/python-formatted-output/)). A specific reason is that I can't easily use thousands separator with the old printf format, while this is trivial with f-strings.

I can rewrite this pretty easily by changing the parameter to format for an f string, [as described here](https://stackoverflow.com/a/54780825).

Can I put in a pull request thingy? Or is there a particular reason for this old format?

**Comment by timhoffm** [comment]
It's the old format because this has already existed for a long time. We cannot simply switch to another formatting mechanism because of backward-compatibility. That would break user code.

I haven't thought this through completely, but I don't think you can use f-strings here because the definition of the string and the formatting are separated, but f-strings are immediately evaluated. It might be possible to use classic `{}`-formatting. But you would have to make sure that `%`-style formatting is still supported and edge-cases (in particular using curly braces and % in the respective other format) are handled correctly.

**Comment by CharlesHe16** [comment]
> I haven't thought this through completely, but I don't think you can use f-strings here because the definition of the string and the formatting are separated, but f-strings are immediately evaluated. 

No, f-string formatting can be separated, just as the fprint format is separated now. Try for yourself:

```
format_f_string = '.2f'

print(f'test string: {42:{format_f_string}}')

format_f_string = '.0f'

print(f'test string: {42:{format_f_string}}')
```

```
>>42.00

>>42
```

> It's the old format because this has already existed for a long time. We cannot simply switch to another formatting mechanism because of backward-compatibility. That would break user code.**

Not trying to argue or even persuade, but genuinely curious: things get deprecated all the time. 

What is the criteria or threshold for changing the format? The fprint or string modulo seems arcane and is discouraged by Python 3 docs. 

>But you would have to make sure that %-style formatting is still supported and edge-cases (in particular using curly braces and % in the respective other format) are handled correctly.

Ok. So I would need to find a way to easily decide whether something is an f-strings format or a string modulo format.

Alternatively, would you accept an additional parameter such as "f_string_format" in addition to "format", or is that too much gore?

**Comment by QuLogic** [comment]
> No, f-string formatting can be separated, just as the fprint format is separated now. Try for yourself:

That example is not strictly equivalent to what's supported in `Sankey`, as it only determines the format of a value, whereas the %-string could contain anything so long as it only interpolated one value. In its usage, the format was only used for determining how a single value was formatted, but there was nothing ensuring that that would be the case, i.e., passing `format='Flow #%d'` was a possibility.

**Comment by CharlesHe16** [comment]
@QuLogic this makes sense.

I guess one solution is to have a second mode of use for when the user enters a tuple. 

The first element of the tuple provides any additional text content, and the second element contains formatting for the f string.

For example `format=('test string %D', '.2f')` with the value of 5 produces `'test string 5.00'`.

This seems awkward and niche.

**Comment by timhoffm** [comment]
> Not trying to argue or even persuade, but genuinely curious: things get deprecated all the time.

> What is the criteria or threshold for changing the format?

For every deprecation we carefully weigh the benefit (clean, intuitive, consistent, ... API) against cost (users have to re-learn, existing code must be changed).

For the sankey format IMHO the benefit is small. It is true that %-formatting has its quirks but it's still a standard and well-known API of Python. And users of `format` will likely need to read the docs anyway to learn about the feature; either way `format` is not canonical enough so that a user could guess what can be passed. OTOH every user of `format` would be affected by the deprecation.

I'm also not clear at what exact API you're aiming. Current API is
~~~
Sankey(..., format='%.2f')
~~~

I could imaging extending to `str.format` (https://docs.python.org/3.9/library/stdtypes.html#str.format) i.e.
~~~
Sankey(..., format='{0:.2f}')
~~~
is valid. That could possibly be checked by `re.match('.*{0(?:.*)}', format)` or similar. That would give you the full flexibility of the [Format String Syntax](https://docs.python.org/3.9/library/string.html#formatstrings).

I don't see how an f-string can be helpful here.

**Comment by CharlesHe16** [comment]
Hi Tim,

>For every deprecation we carefully weigh the benefit (clean, intuitive, consistent, ... API) against cost (users have to re-learn, existing code must be changed).

>For the sankey format IMHO the benefit is small. It is true that %-formatting has its quirks but it's still a standard and well-known API of Python. And users of format will likely need to read the docs anyway to learn about the feature; either way format is not canonical enough so that a user could guess what can be passed. OTOH every user of format would be affected by the deprecation.

This makes sense, thanks for explaining with as much precision as you did. 

>I don't see how an f-string can be helpful here.

I expect you're right and I will follow this.

Basically, f-strings are my way of putting numbers in strings. I've used str.format and also %-formatting in the past, and to me, f-string seems to dominate these other methods. 

In fact, until reading your message, I believed that f-strings are the one best canonical way of putting numbers in strings, so I planned to "upgrade" `format` with them.

Zooming out, I agree this seems niche. 

I am willing to write a pull request to use str.format if you explicitly say this is useful, otherwise I will just monkey patch or move on.

**Comment by WeatherGod** [comment]
Just a quibble on an assertion made earlier in the thread. %-style
formatting is *not* obsolete or discouraged, per se. It is a perfectly fine
way of doing things. str.format() (and by extension, f-strings) are
powerful in other ways. And don't forget about the lowly string templating
module! I have actually used these three features together a few times in
nested configuration systems for some projects.

It isn't going away any time soon.

**Comment by CharlesHe16** [comment]
>Just a quibble on an assertion made earlier in the thread. %-style
formatting is *not* obsolete or discouraged, per se. It is a perfectly fine
way of doing things

Can you give an example of where this is better than f-strings, preferably one that isn't stylistic?

**Comment by WeatherGod** [comment]
I never claimed it was better. I just said that it was a perfectly fine way
of doing things and not obsolete. There isn't anything inherently wrong
with it. What is nice is its symplicity, but that is a stylistic concern.
As far as I am aware, the python devs have given no signal that %-style
formatting will ever be removed.

**Comment by CharlesHe16** [comment]
People can have opinions and prefer different ways of doing things. Still, having a single way of doing things is an advantage and personal preference doesn't seem to be a good argument.

You don't have a reason as to why %-formatting is better, and seem to agree it may be no better. 

On the other hand, I can present reasons why f-strings are better (they are described in the PEP)

>I have actually used these three features together a few times in
nested configuration systems for some projects.

It seems objectively worse to mix three different formats. 

Arguing about this from an ego or preference point of view seems very unpromising and I don't seek to do this. 

I want to understand your opinions if there is content that makes anything of what I said above incorrect.

**Comment by jklymak** [comment]
C-style string formatting is immediately familiar to anyone who used most other programming languages.  F strings, not so much.

**Comment by WeatherGod** [comment]
Please stop reading more into what I have written than is actually there.
There is no ego being expressed or preference being stated. I explicitly
said at the beginning that I had a quibble with your claim that the python
documentation made it seem like %-formatting was "obsolete and
discouraged". This is *not* the case. That is all I was saying! Please do
not continue this false narrative.

I like str.format()-style formatting for many reasons, particularly the
ability to create custom format notations, al la datetime objects. I
actually have no objection with extending sanskey somehow to also include
other formatting methods. Perhaps we could utilize the tick formatter
infrastructure to do this?

The only reason why I said anything in this conversation is to dispel the
myth that %-formatting is somehow obsolete and discouraged in python (your
words). In fact, the documentation makes it clear that there are times when
having both types of formatting available is preferable (I don't have a
reference on-hand, unfortunately). "Obsolete and discouraged" implies a
movement towards deprecation and removal from python, which is *not* the
case.

**Comment by tacaswell** [comment]
Can we put a lid on the relative merits of the string formatting styles?  They all exist in the Python language which is something we have to support.   If you want to have a discussion about changing that or the relative strengths, please take that to python-ideas@python.org ;).

----

If we were going to extend this API to support new-style format strings I think we should go the route (riffing on some of Antony's comments else where) of accepting either a string (which is formatted with `%`) or a callable that accepts a number and returns a string (so `'{.2f}'.format` would be a valid input).  We don't pick up coupled kwargs and there is a reliable way to tell which input the user gave us, and we maintain full back-compatibility.  

It also gives the user far _more_ power as now they can inject an arbitrary function and, for example, map number to emoji.

**Comment by tacaswell** [comment]
If folks agree with the `str` or `callable` path, I think this is a good first issue as the API extension is narrow and self-contained to the sankey code.  It will need docs, tests, an example, and a whats-new entry.

**Comment by ianhi** [comment]
> self-contained to the sankey code.

There are a few other places in matplotlib that have similar string formatting. It would be nice if a change to accept both types of formatting was applied uniformly across them. Most notable is probably the Slider widgets which also use `%` style formatting. grepping around I also found 
https://github.com/matplotlib/matplotlib/blob/67aa50d164669a1340fe05e06dfc7223e25c87b3/lib/matplotlib/contour.py#L918-L919

which only allows for callables. It would be nice these were all consistent and all allowed for both styles.

**Comment by jklymak** [comment]
I think `clabel` should probably be the gold-standard here?  https://matplotlib.org/3.3.0/api/contour_api.html#matplotlib.contour.ContourLabeler.clabel

```
        fmt : `.Formatter` or str or callable or dict, optional
            How the levels are formatted:

            - If a `.Formatter`, it is used to format all levels at once, using
              its `.Formatter.format_ticks` method.
            - If a str, it is interpreted as a %-style format string.
            - If a callable, it is called with one level at a time and should
              return the corresponding label.
            - If a dict, it should directly map levels to labels.

            The default is to use a standard `.ScalarFormatter`.
```

So maybe what `clabel` does should be made a `_api` helper and used elsewhere?   (basically `ContourLabeller.get_text`)

**Comment by tacaswell** [comment]
I have a small concern about the Formatter and dict layers.  The first seems tuned to clabel needing many strings (not just one) and the second seems a shortcut to `fmt=lambda x: my_dict[x]`.

**Comment by timhoffm** [comment]
Agree that `Formatter` and `dict` are not very useful in general, and in particular for Sankey.

Also, with str and callable, you can do everything.

**Comment by CharlesHe16** [comment]
I think I understand @tacaswell's solution with the `str or callable` route. This makes sense.

Plus, we could then measure flows in nyan cats: 

<img src="https://user-images.githubusercontent.com/23063929/102678266-29dbcf00-415c-11eb-967a-c3f7d8a15164.png" width="80"><img src="https://user-images.githubusercontent.com/23063929/102678266-29dbcf00-415c-11eb-967a-c3f7d8a15164.png" width="80">

I see the issue is in "unassigned" and "needs sorting". Some issues have been in those categories for a long time. 

What's the next step?

**Comment by timhoffm** [comment]
**The desired API is deciced based on the above discussion:** In addition to the existing string, the `format` parameter should support a callable.

**Next step:** Someone needs to open a pull request with the desired change.

I'm labeling this as "Good first issue" because the extension is straight forward, and you can take

- https://matplotlib.org/devdocs/api/contour_api.html#matplotlib.contour.ContourLabeler.clabel `fmt` or
- https://matplotlib.org/devdocs/api/_as_gen/matplotlib.pyplot.pie.html#matplotlib.pyplot.pie `autopct`

as a template.

**Comment by tacaswell** [comment]
We do not use the "assign" feature of the GH (it makes lots of sense in an organization where everyone working on the code is an employee in a reporting chain and part of the org, it makes less sense in a community project where most contributors are volunteers and many are not in the org).

I have re-milestoned this to 3.5 (the next-next feature release) because we are aiming to have 3.4 out "soon", this is not something we should block 3.4 on and work has not started yet.

@CharlesHe16 If you are interested in working on this that would be great :)

**Comment by CharlesHe16** [comment]
Ok, I did something! I've created a [pull request, 19187](https://github.com/matplotlib/matplotlib/pull/19187).

By the way, as the [Contributor part of this Guide put it](https://matplotlib.org/devdocs/devel/contributing.html), I "feel insecure about it". 

If anyone has a little time, it would be helpful to point out any errors I've made, either in content or procedure, and briefly suggest how to address it.

Thanks!

**Comment by timhoffm** [comment]
Ping @tacaswell @jklymak: Please add @CharlesHe16 to the `#incubator` channel if not already done so.

**Comment by jklymak** [comment]
They would have to add themselves to gitter first I think

**Comment by CharlesHe16** [comment]
Ok, I don't know what being added to the `#incubator` channel means, but I'm all for it!

> They would have to add themselves to gitter first I think

I'm not familiar with Gitter, but I think I signed up using my Github account. 

I cannot find a way to get a "profile page" on Gitter, but I think I show up on searches, producing this URL: https://gitter.im/CharlesHe16

My Gitter user id might be: `5ff29b01d73408ce4ff7ea92 `

**Comment by tacaswell** [comment]
Closed by #19187
